### PR TITLE
Multi label models

### PR DESCRIPTION
**Pull request by RichJackson**
reworked a bunch of logic to clean up TransformersModelForTokenClassificationNerStep ahead of multilabel token classification. Removed some dead code for obsolete concepts.

See commit comments for specific changes

### CODE REVIEW

File: kazu/steps/ner/tokenized_word_processor.py [hunk 1]
```diff
@@ -224,81 +204,98 @@ def process_next_word(self, word: TokenizedWord) -> None:
...
+        """
+        for word in words:
+            self.process_next_word(word)
+        for clazz in set(self.active_labels):
+            self.close_spans(clazz)
+        return self.closed_spans
+
+    def get_class_labels(self, word: TokenizedWord) -> set[str]:
+        class_labels: set[str] = set()
+        indices_of_confs_above_threshold = torch.argwhere(word.token_confidences > 0)
```
Comment (wonjininfo) [inline kazu/steps/ner/tokenized_word_processor.py:273]: Hi, I just came across this line—does `word.token_confidences` represent logits? 
Previously, the threshold was 0.5, suggesting it was a probability (between 0 and 1). However, based on this commit (`torch.argwhere(word.token_confidences > 0)`), it seems it can have negative values, so I guess it represents logits in the current implementation.

It's been a while since I last looked at this, so I might have misunderstood the code. 😅 Please ignore if that's the case!

Comment (RichJackson) [reply kazu/steps/ner/tokenized_word_processor.py:273]: Hi @wonjininfo ! Thanks for the check - actually, I've changed your implementation slightly - token confidences are now either 0 or 1. I should probably change these variable names as they're confusing

Comment (wonjininfo) [reply kazu/steps/ner/tokenized_word_processor.py:273]: Thanks for the explanation! These lines look good to me

File: kazu/steps/ner/tokenized_word_processor.py [hunk 2]
```diff
@@ -142,38 +147,32 @@ def close_spans(self):
...
 
         :param word:
+        :param bio_and_class_labels:
         :return:
         """
-        pass
+        classes_set = set(x[1] for x in bio_and_class_labels)
+        if None in classes_set or self.text[word.word_char_start - 1] in self.span_breaking_chars:
+            return False
+        else:
```
Comment (paluchasz) [inline kazu/steps/ner/tokenized_word_processor.py:166]: nit: else not needed

File: kazu/steps/ner/tokenized_word_processor.py [hunk 3]
```diff
@@ -224,81 +204,98 @@ def process_next_word(self, word: TokenizedWord) -> None:
...
-        classes_set.discard(None)
-        if (
-            self.text[word.word_char_start - 1] not in self.span_breaking_chars
-            or len(classes_set) > 0
-        ):
-            return True
-        else:
+        if not class_labels or self.text[word.word_char_start - 1] in self.span_breaking_chars:
             return False
+        else:
```
Comment (paluchasz) [inline kazu/steps/ner/tokenized_word_processor.py:297]: nit: else not needed

File: kazu/steps/ner/tokenized_word_processor.py [hunk 4]
```diff
@@ -333,56 +331,37 @@ class to find entity spans according to their internal logic
...
 
+    def _make_multilabel_span_finder(self, text: str) -> MultilabelSpanFinder:
+        return MultilabelSpanFinder(text=text, id2label=self.id2label)
+
     def make_span_finder(self, text: str) -> SpanFinder:
-        if self.detect_subspans:
-            return self._make_smart_span_finder(text)
+        if self.use_multilabel:
+            return self._make_multilabel_span_finder(text)
         else:
```
Comment (paluchasz) [inline kazu/steps/ner/tokenized_word_processor.py:365]: nit: else not needed

File: kazu/steps/ner/tokenized_word_processor.py [hunk 5]
```diff
@@ -37,18 +38,40 @@ class TokWordSpan:
...
     #: entity_class
     clazz: str
-    #: is this a subspan? only relevant if using SmartSpanFinder
-    subspan: Optional[bool] = None
     #: words associated with this span
     tok_words: list[TokenizedWord] = field(default_factory=list)
 
 
 class SpanFinder(ABC):
+    def __init__(self, text: str, id2label: dict[int, str]):
```
Comment (paluchasz) [inline kazu/steps/ner/tokenized_word_processor.py:46]: nit: `-> None`

File: kazu/steps/ner/hf_token_classification.py [hunk 6]
```diff
@@ -67,24 +62,25 @@ def __init__(
...
         stride: int,
         max_sequence_length: int,
-        labels: list[str],
-        detect_subspans: bool = False,
-        threshold: Optional[float] = None,
+        tokenized_word_processor: TokenizedWordProcessor,
+        keys_to_use: Iterable[str],
         entity_splitter: Optional[NonContiguousEntitySplitter] = None,
-        strip_re: Optional[dict[str, str]] = None,
+        device: Optional[str] = None,
```
Comment (paluchasz) [inline kazu/steps/ner/hf_token_classification.py:68]: should the default be `None` or better `cpu`? Or even better it can automatically detect if cuda is available (`"cuda" if torch.cuda.is_available() else "cpu" as a global variable)` )? 
The doc string says defaults to cpu.

**Review by paluchasz — APPROVED** [review]
Looks good, just some nits here and there

File: kazu/steps/ner/hf_token_classification.py (hunk 6 above)
Comment (RichJackson) [reply kazu/steps/ner/hf_token_classification.py:68]: good point this should be cpu. I don't want to make it automatically detect/use the GPU, as this might have unexpected consequences for the user. For instance, in an environment where the GPU is doing something else, and the user has Kazu running in the background, it might cause all kinds of problems
